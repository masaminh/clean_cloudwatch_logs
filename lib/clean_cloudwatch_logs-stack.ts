import * as cdk from 'aws-cdk-lib';
import {
  aws_stepfunctions as sfn,
  aws_stepfunctions_tasks as tasks,
  aws_events as events,
  aws_events_targets as targets,
  aws_lambda as lambda,
  aws_lambda_nodejs as lambdaNodejs,
  aws_iam as iam,
  aws_sqs as sqs,
} from 'aws-cdk-lib';
import * as construct from 'constructs';

const retryProps: sfn.RetryProps = {
  errors: ['CloudWatchLogs.CloudWatchLogsException', 'CloudWatchLogs.SdkClientException'],
  interval: cdk.Duration.seconds(1),
  maxAttempts: 5,
  backoffRate: 2,
};

interface StackProps extends cdk.StackProps {
  queueArn: string;
}

// eslint-disable-next-line import/prefer-default-export
export class CleanCloudwatchLogsStack extends cdk.Stack {
  constructor(scope: construct.Construct, id: string, props: StackProps) {
    super(scope, id, props);

    const getLogGroupsLambda = new lambdaNodejs.NodejsFunction(this, 'get_log_groups', {
      architecture: lambda.Architecture.ARM_64,
      timeout: cdk.Duration.minutes(1),
      tracing: lambda.Tracing.ACTIVE,
      runtime: lambda.Runtime.NODEJS_20_X,
    });

    getLogGroupsLambda.addToRolePolicy(new iam.PolicyStatement({
      resources: ['arn:aws:logs:*:*:log-group:*:*'],
      actions: ['logs:DescribeLogGroups'],
    }));

    const getLogGroups = new tasks.LambdaInvoke(this, 'GetLogGroups', {
      lambdaFunction: getLogGroupsLambda,
      inputPath: '$.time',
      payloadResponseOnly: true,
    });

    const getLogStreamsLambda = new lambdaNodejs.NodejsFunction(this, 'get_log_streams', {
      architecture: lambda.Architecture.ARM_64,
      timeout: cdk.Duration.minutes(5),
      tracing: lambda.Tracing.ACTIVE,
      runtime: lambda.Runtime.NODEJS_20_X,
    });

    getLogStreamsLambda.addToRolePolicy(new iam.PolicyStatement({
      resources: ['arn:aws:logs:*:*:log-group:*:log-stream:*'],
      actions: ['logs:DescribeLogStreams'],
    }));

    const getLogStreams = new tasks.LambdaInvoke(this, 'GetLogStreams', {
      lambdaFunction: getLogStreamsLambda,
      payloadResponseOnly: true,
    });
    const deleteLogStreamsLambda = new lambdaNodejs.NodejsFunction(this, 'delete_log_streams', {
      architecture: lambda.Architecture.ARM_64,
      timeout: cdk.Duration.minutes(15),
      tracing: lambda.Tracing.ACTIVE,
      runtime: lambda.Runtime.NODEJS_20_X,
    });

    deleteLogStreamsLambda.addToRolePolicy(new iam.PolicyStatement({
      resources: ['arn:aws:logs:*:*:log-group:*:log-stream:*'],
      actions: ['logs:DeleteLogStream'],
    }));

    const deleteLogStreams = new tasks.LambdaInvoke(this, 'DeleteLogStreams', {
      lambdaFunction: deleteLogStreamsLambda,
      payloadResponseOnly: true,
    });

    const notifyMessage = new tasks.EvaluateExpression(this, 'NotifyMessage', {
      // eslint-disable-next-line no-template-curly-in-string
      expression: '`Region: ${$.region}, Loggroup: ${$.logGroupName} is empty.`',
      resultPath: '$.notifyMessage',
    });

    const notifyEmptyLogGroup = new tasks.SqsSendMessage(this, 'NotifyEmptyLogGroup', {
      queue: sqs.Queue.fromQueueArn(this, 'NotifyQueue', props.queueArn),
      messageBody: sfn.TaskInput.fromObject({
        webhookname: 'Develop',
        message: sfn.JsonPath.stringAt('$.notifyMessage'),
      }),
    });

    const getLogStreamsMap = new sfn.Map(this, 'GetLogStreamsMap', {
      itemsPath: sfn.JsonPath.stringAt('$.targetLogGroups'),
      itemSelector: {
        'eventTime.$': '$.eventTime',
        'logGroupInfo.$': '$$.Map.Item.Value',
      },
      maxConcurrency: 1,
      resultPath: sfn.JsonPath.DISCARD,
    });

    const setRetention = new tasks.CallAwsService(this, 'SetRetentionInDays', {
      service: 'cloudwatchlogs',
      action: 'putRetentionPolicy',
      iamAction: 'logs:putRetentionPolicy',
      iamResources: ['arn:aws:logs:*:*:log-group:*:*'],
      resultPath: sfn.JsonPath.DISCARD,
      parameters: {
        'LogGroupName.$': '$.logGroupInfo.name',
        RetentionInDays: 30,
      },
    }).addRetry(retryProps);

    const setRetentionMap = new sfn.Map(this, 'SetRetentionMap', {
      itemsPath: sfn.JsonPath.stringAt('$.noRetentionLogGroups'),
      itemSelector: {
        'logGroupInfo.$': '$$.Map.Item.Value',
      },
      maxConcurrency: 1,
      resultPath: sfn.JsonPath.DISCARD,
    });

    const parallelLogGroups = new sfn.Parallel(this, 'ParallelLogGroups');

    parallelLogGroups.branch(
      getLogStreamsMap.itemProcessor(getLogStreams.next(
        new sfn.Choice(this, 'IsEmptyLogGroup')
          .when(
            sfn.Condition.booleanEquals('$.isEmpty', true),
            notifyMessage.next(notifyEmptyLogGroup),
          )
          .otherwise(deleteLogStreams),
      )),
      setRetentionMap.itemProcessor(setRetention),
    );

    const stateMachine = new sfn.StateMachine(this, 'StateMachine', {
      definitionBody: sfn.DefinitionBody.fromChainable(getLogGroups.next(parallelLogGroups)),
    });

    // eslint-disable-next-line no-new
    new events.Rule(this, 'Rule', {
      schedule: events.Schedule.rate(cdk.Duration.days(1)),
      targets: [new targets.SfnStateMachine(stateMachine)],
      enabled: true,
    });
  }
}
