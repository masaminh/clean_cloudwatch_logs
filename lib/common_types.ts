import { isObject, isString, isNumber } from './type_utils';

export type LogGroupInfoType = {
  name: string;
  retentionInDays?: number;
  region: string;
}

export function isLogGroupInfoType(arg: unknown): arg is LogGroupInfoType {
  return (
    isObject<LogGroupInfoType>(arg)
    && isString(arg.name)
    && (arg.retentionInDays == null || isNumber(arg.retentionInDays))
    && isString(arg.region)
  );
}

export type LogStreamInfoType = {
  name: string;
}

export function isLogStreamInfoType(arg: unknown): arg is LogStreamInfoType {
  return (
    isObject<LogStreamInfoType>(arg)
    && isString(arg.name)
  );
}
