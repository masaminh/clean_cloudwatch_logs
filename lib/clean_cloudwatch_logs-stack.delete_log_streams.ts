import * as AWSXRay from 'aws-xray-sdk';
import {
  CloudWatchLogsClient, DeleteLogStreamCommand,
} from '@aws-sdk/client-cloudwatch-logs';
import { LogStreamInfoType, isLogStreamInfoType } from './common_types';

type InputType = {
  logGroupName: string,
  targetLogStreams: LogStreamInfoType[];
}

type OutputType = {}

const client = AWSXRay.captureAWSv3Client(
  new CloudWatchLogsClient({
    maxAttempts: 10,
  }),
);

function isInputType(arg: any): arg is InputType {
  return (
    arg != null
    && typeof arg.logGroupName === 'string'
    && Array.isArray(arg.targetLogStreams)
    && arg.targetLogStreams.every(isLogStreamInfoType)
  );
}

// eslint-disable-next-line import/prefer-default-export
export async function handler(input: any): Promise<OutputType> {
  const startTime = Date.now();

  if (!isInputType(input)) {
    return {};
  }

  // eslint-disable-next-line no-restricted-syntax
  for (const logStreamInfo of input.targetLogStreams) {
    // eslint-disable-next-line no-await-in-loop
    await client.send(new DeleteLogStreamCommand({
      logGroupName: input.logGroupName,
      logStreamName: logStreamInfo.name,
    }));

    if (Date.now() - startTime > 10 * 60 * 1000) {
      break;
    }
  }

  return {};
}
