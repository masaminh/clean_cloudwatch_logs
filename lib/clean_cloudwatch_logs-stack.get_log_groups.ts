import { CloudWatchLogsClient, paginateDescribeLogGroups } from '@aws-sdk/client-cloudwatch-logs';
import { LogGroupInfoType } from './common_types';
import { isString } from './type_utils';

type OutputType = {
  eventTime: number,
  targetLogGroups: LogGroupInfoType[],
  noRetentionLogGroups: LogGroupInfoType[],
  emptyLogGroups: LogGroupInfoType[],
}

const client = new CloudWatchLogsClient({ maxAttempts: 5 });

// eslint-disable-next-line import/prefer-default-export
export async function handler(time: unknown): Promise<OutputType> {
  if (!isString(time)) {
    return {
      eventTime: 0,
      targetLogGroups: [],
      noRetentionLogGroups: [],
      emptyLogGroups: [],
    };
  }

  const eventTime = Date.parse(time);
  const thresholdTimeEpoch = eventTime - 7 * 24 * 60 * 60 * 1000;
  const targetLogGroups: LogGroupInfoType[] = [];
  const noRetentionLogGroups: LogGroupInfoType[] = [];
  const emptyLogGroups: LogGroupInfoType[] = [];

  // eslint-disable-next-line no-restricted-syntax
  for await (const page of paginateDescribeLogGroups({ client }, {})) {
    page.logGroups?.forEach((logGroup) => {
      if (logGroup.logGroupName == null) {
        return;
      }

      if ((logGroup.creationTime ?? Number.MAX_SAFE_INTEGER) > thresholdTimeEpoch) {
        return;
      }

      let logGroups: LogGroupInfoType[];

      const logGroupInfo: LogGroupInfoType = {
        name: logGroup.logGroupName,
        retentionInDays: logGroup.retentionInDays,
      };

      if (logGroup.retentionInDays == null) {
        logGroups = noRetentionLogGroups;
      } else if (logGroup.storedBytes === 0) {
        logGroups = emptyLogGroups;
      } else {
        logGroups = targetLogGroups;
      }

      logGroups.push(logGroupInfo);
    });
  }

  return {
    eventTime,
    targetLogGroups,
    noRetentionLogGroups,
    emptyLogGroups,
  };
}
