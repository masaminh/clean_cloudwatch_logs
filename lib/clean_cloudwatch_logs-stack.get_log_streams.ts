import * as AWSXRay from 'aws-xray-sdk';
import { CloudWatchLogsClient, paginateDescribeLogStreams } from '@aws-sdk/client-cloudwatch-logs';
import { LogStreamInfoType, isLogGroupInfoType, LogGroupInfoType } from './common_types';

type InputType = {
  eventTime: number;
  logGroupInfo: LogGroupInfoType;
}

type OutputType = {
  logGroupName: string,
  targetLogStreams: LogStreamInfoType[];
}

const client = AWSXRay.captureAWSv3Client(
  new CloudWatchLogsClient({
    maxAttempts: 10,
  }),
);

function isInputType(arg: any): arg is InputType {
  return (
    arg != null
    && typeof arg.eventTime === 'number'
    && isLogGroupInfoType(arg.logGroupInfo)
  );
}

// eslint-disable-next-line import/prefer-default-export
export async function handler(input: any): Promise<OutputType> {
  if (!isInputType(input)) {
    return {
      logGroupName: '',
      targetLogStreams: [],
    };
  }

  const { logGroupInfo } = input;

  const logGroupName = logGroupInfo.name;
  const { retentionInDays } = logGroupInfo;

  if (retentionInDays == null) {
    return {
      logGroupName: '',
      targetLogStreams: [],
    };
  }

  const targetLogStreams: LogStreamInfoType[] = [];

  // eslint-disable-next-line no-restricted-syntax
  for await (const page of paginateDescribeLogStreams({ client }, {
    logGroupName,
  })) {
    page.logStreams?.forEach((logStream) => {
      if (logStream.logStreamName == null) {
        return;
      }

      if (logStream.lastIngestionTime == null) {
        return;
      }

      const threshold = input.eventTime - retentionInDays * 24 * 60 * 60 * 1000;

      if (logStream.lastIngestionTime > threshold) {
        return;
      }

      targetLogStreams.push({
        name: logStream.logStreamName,
      });
    });
  }

  return {
    logGroupName,
    targetLogStreams,
  };
}
