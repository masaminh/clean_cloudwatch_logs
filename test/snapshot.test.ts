import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CleanCloudwatchLogsStack } from '../lib/clean_cloudwatch_logs-stack';

export const ignoreAssetHashSerializer = {
  test: (val: unknown) => typeof val === 'string',
  serialize: (val: string) => {
    return `"${val.replace(/([A-Fa-f0-9]{64}.zip)/, 'HASH-REPLACED.zip')}"`;
  },
};

test('snapshot test', () => {
  const app = new cdk.App();
  const queueArn = 'arn:aws:sqs:ap-northeast-1:123456789012:Queue';
  console.log(queueArn);

  const stack = new CleanCloudwatchLogsStack(app, 'MyTestStack',  {
    env: { region: 'ap-northeast-1' }, queueArn,
  });
  // スタックからテンプレート(JSON)を生成
  const template = Template.fromStack(stack).toJSON();

  expect.addSnapshotSerializer(ignoreAssetHashSerializer);

  // 生成したテンプレートとスナップショットが同じか検証
  expect(template).toMatchSnapshot();
});
