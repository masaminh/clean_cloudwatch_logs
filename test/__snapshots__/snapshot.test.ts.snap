// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Eval41256dc5445742738ed917bc818694e54EB1134F": {
      "DependsOn": [
        "Eval41256dc5445742738ed917bc818694e5ServiceRoleA1AB6027",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "Eval41256dc5445742738ed917bc818694e5ServiceRoleA1AB6027",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "Eval41256dc5445742738ed917bc818694e5ServiceRoleA1AB6027": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Rule4C995B7F": {
      "Properties": {
        "ScheduleExpression": "rate(1 day)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Ref": "StateMachine2E01A3A5",
            },
            "Id": "Target0",
            "RoleArn": {
              "Fn::GetAtt": [
                "StateMachineEventsRoleDBCDECD1",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "StateMachine2E01A3A5": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"GetLogGroups","States":{"GetLogGroups":{"Next":"ParallelLogGroups","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$.time","Resource":"",
              {
                "Fn::GetAtt": [
                  "getloggroupsD7CF3457",
                  "Arn",
                ],
              },
              ""},"ParallelLogGroups":{"Type":"Parallel","End":true,"Branches":[{"StartAt":"GetLogStreamsMap","States":{"GetLogStreamsMap":{"Type":"Map","ResultPath":null,"End":true,"ItemsPath":"$.targetLogGroups","ItemSelector":{"eventTime.$":"$.eventTime","logGroupInfo.$":"$$.Map.Item.Value"},"ItemProcessor":{"ProcessorConfig":{"Mode":"INLINE"},"StartAt":"GetLogStreams","States":{"GetLogStreams":{"Next":"IsEmptyLogGroup","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Resource":"",
              {
                "Fn::GetAtt": [
                  "getlogstreamsD20145F9",
                  "Arn",
                ],
              },
              ""},"IsEmptyLogGroup":{"Type":"Choice","Choices":[{"Variable":"$.isEmpty","BooleanEquals":true,"Next":"NotifyMessage"}],"Default":"DeleteLogStreams"},"DeleteLogStreams":{"End":true,"Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Resource":"",
              {
                "Fn::GetAtt": [
                  "deletelogstreams9F60BE28",
                  "Arn",
                ],
              },
              ""},"NotifyMessage":{"Next":"NotifyEmptyLogGroup","Type":"Task","ResultPath":"$.notifyMessage","Resource":"",
              {
                "Fn::GetAtt": [
                  "Eval41256dc5445742738ed917bc818694e54EB1134F",
                  "Arn",
                ],
              },
              "","Parameters":{"expression":"\`Region: \${$.region}, Loggroup: \${$.logGroupName} is empty.\`","expressionAttributeValues":{"$.region.$":"$.region","$.logGroupName.$":"$.logGroupName"}}},"NotifyEmptyLogGroup":{"End":true,"Type":"Task","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::sqs:sendMessage","Parameters":{"QueueUrl":"https://sqs.ap-northeast-1.",
              {
                "Ref": "AWS::URLSuffix",
              },
              "/123456789012/Queue","MessageBody":{"webhookname":"Develop","message.$":"$.notifyMessage"}}}}},"MaxConcurrency":1}}},{"StartAt":"SetRetentionMap","States":{"SetRetentionMap":{"Type":"Map","ResultPath":null,"End":true,"ItemsPath":"$.noRetentionLogGroups","ItemSelector":{"logGroupInfo.$":"$$.Map.Item.Value"},"ItemProcessor":{"ProcessorConfig":{"Mode":"INLINE"},"StartAt":"SetRetentionInDays","States":{"SetRetentionInDays":{"End":true,"Retry":[{"ErrorEquals":["CloudWatchLogs.CloudWatchLogsException","CloudWatchLogs.SdkClientException"],"IntervalSeconds":1,"MaxAttempts":5,"BackoffRate":2}],"Type":"Task","ResultPath":null,"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:cloudwatchlogs:putRetentionPolicy","Parameters":{"LogGroupName.$":"$.logGroupInfo.name","RetentionInDays":30}}}},"MaxConcurrency":1}}}]}}}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "StateMachineEventsRoleDBCDECD1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineEventsRoleDefaultPolicyFB602CA9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "StateMachine2E01A3A5",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineEventsRoleDefaultPolicyFB602CA9",
        "Roles": [
          {
            "Ref": "StateMachineEventsRoleDBCDECD1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StateMachineRoleB840431D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "states.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "getloggroupsD7CF3457",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "getloggroupsD7CF3457",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "getlogstreamsD20145F9",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "getlogstreamsD20145F9",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "deletelogstreams9F60BE28",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "deletelogstreams9F60BE28",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Eval41256dc5445742738ed917bc818694e54EB1134F",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Eval41256dc5445742738ed917bc818694e54EB1134F",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "sqs:SendMessage",
              "Effect": "Allow",
              "Resource": "arn:aws:sqs:ap-northeast-1:123456789012:Queue",
            },
            {
              "Action": "logs:putRetentionPolicy",
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:log-group:*:*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
        "Roles": [
          {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "deletelogstreams9F60BE28": {
      "DependsOn": [
        "deletelogstreamsServiceRoleDefaultPolicy706D0026",
        "deletelogstreamsServiceRole859AA330",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "deletelogstreamsServiceRole859AA330",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "deletelogstreamsServiceRole859AA330": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "deletelogstreamsServiceRoleDefaultPolicy706D0026": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "logs:DeleteLogStream",
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:log-group:*:log-stream:*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "deletelogstreamsServiceRoleDefaultPolicy706D0026",
        "Roles": [
          {
            "Ref": "deletelogstreamsServiceRole859AA330",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "getloggroupsD7CF3457": {
      "DependsOn": [
        "getloggroupsServiceRoleDefaultPolicy61B31DEF",
        "getloggroupsServiceRole2F7300D4",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "getloggroupsServiceRole2F7300D4",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 60,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "getloggroupsServiceRole2F7300D4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "getloggroupsServiceRoleDefaultPolicy61B31DEF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "logs:DescribeLogGroups",
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:log-group:*:*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "getloggroupsServiceRoleDefaultPolicy61B31DEF",
        "Roles": [
          {
            "Ref": "getloggroupsServiceRole2F7300D4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "getlogstreamsD20145F9": {
      "DependsOn": [
        "getlogstreamsServiceRoleDefaultPolicy10C98299",
        "getlogstreamsServiceRoleF2C01167",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "getlogstreamsServiceRoleF2C01167",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 300,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "getlogstreamsServiceRoleDefaultPolicy10C98299": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "logs:DescribeLogStreams",
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:log-group:*:log-stream:*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "getlogstreamsServiceRoleDefaultPolicy10C98299",
        "Roles": [
          {
            "Ref": "getlogstreamsServiceRoleF2C01167",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "getlogstreamsServiceRoleF2C01167": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
